docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,705 @@
+  workflow is stalled or has incomplete tasks
+  - similarly for stopped workflows (also: "not started", "ran to completion",
+    "died", etc.).
+  - we need a mechanism to ensure that any workflows that stopped unexpectedly
+    don't get missed and that users can easily find their most recent workflows
+    (e.g. `cylc review` orders by last activity).
+
+## New Event Handler
+
+We should add a new **incomplete task** event that is triggered whenever a task
+finishes without completing all of its expected outputs.
+
+## Runahead limit
+
+See https://github.com/cylc/cylc-flow/issues/4258. Some users understand
+runahead limiting as preventing the fastest tasks from getting too far ahead of
+the slowest.
+
+Incomplete tasks and partially satisfied prerequisites both represent tasks
+that will, or at least might after user intervention, still run at their cycle
+point, so they should count towards the runahead limit.
+
+(They should not count toward an *active cycle point* based limit, if we decide
+to implement that).
+
+Note that future triggers have the potential to cause complications with the
+runahead limit by spawning tasks at old cycles. This kind of issue would have
+been obvious at Cylc 7 because the task waiting for the future trigger would
+already have been in the task pool.
+
+## Cylc 7 compatibility
+
+Cylc 7 workflows will not be using the new syntax which means they will only
+have expected outputs (but that's OK - Cylc 7 outputs **are** "expected" as
+explained above).
+
+They also use suicide triggers to remove:
+- pre-spawned waiting tasks that would never run (i.e. from unused alternate
+  branches)
+- failed tasks that don't need to be retained in the pool
+
+For example, in this Cylc 8 graph the `c1 => c2` branch does not get spawned
+at all unless `a` fails:
+```
+a? => b1 => b2
+a?fail => c1 => c2
+b2 | c2 => d
+```
+
+The Cylc 7 version of this already works out of the box on Cylc 8 master, for
+backward compatibility:
+```
+a => b1 => b2
+a:fail => c1 => c2
+b2 | c2 => d
+a => !c1 & !c2
+a:fail => !a & !b1 & !b2
+d => !a # prevent "a" disappearing from the GUI too quickly
+```
+
+This works because Cylc 8 suicide triggers spawn (or update) children just like
+normal triggers. Once all of a task's suicide prerequisites are satisfied it
+is removed immediately. A task that has a single suicide trigger and would not
+otherwise be spawned in Cylc 8 will have a transient presence in the task pool
+because the suicide trigger that spawns it also satisfies its only suicide
+prerequisite.
+
+It doesn't matter if a task still has unsatisfied normal prerequisites when it
+suicides. The other triggers can't spawn it again in the same flow thanks to
+our conditional reflow prevention mechanism.
+
+For the Cylc 7 graph above to work with the new optional output syntax we just
+need to infer that `a:succeed` and `a:fail` are optional, to avoid `a` ending
+up marked as incomplete both ways.
```

Comment:
I'm not sure what you mean here. My assumption is that we have to treat `succeed` and `fail` as required so that `a` remains in the task pool until such time as it is removed via a suicide trigger.

Reply:
No, I don't think so.  Suicide triggers necessarily work differently under spawn-on-demand, but that doesn't matter so long as we get the same outcome.  Here's a simpler example:

```
# Native Cylc 8
a? => b
a?fail => c
```
... `b` only gets spawned if `a` succeeds; and `c` only gets spawned if `a` fails

```
# Cylc 7
a => b
a:fail => c
a:fail => !b
a => !c
c => !a 
```
...both `b` and `c` are pre-spawned. If `a` fails, it gets removed later on by suicide. And `b` or `c` get removed immediately, depending on the final state of `a`.

If we run the Cylc 7 case with Cylc 8:
- we can safely infer that success (and failure) of `a` is optional, because the alternative is impossible
   - then we don't end up with `a` incomplete, whichever outcome occurs
- later, if `a` fails (say), the suicide triggers will:
- spawn `b`, which will immediately be removed again because it's suicide prerequisite is satisfied by the trigger that spawned it
-  be blocked from spawning `a` because `a` was already spawned earlier in this flow

Reply:
What you're suggesting probably is an option too, because the suicide trigger would still remove if the incomplete task.  But it would be better to avoid the incomplete task in the first place, if we can.

Reply:
Even @oliver-sanders nasty Cylc 7 aborted pipeline case above runs correctly on current master:
```
a1 => b1
a1:fail | b1 => c => d1
a1:fail & c => !a1 & !b1 & !d1
a2 => b2
a2:fail | b2 => c => d2
a2:fail & c => !a2 & !b2 & !d2
```
Result with `a2` failing:
```console
$ cylc cat-log d8/run1 | grep trig
2021-07-29T17:19:07+12:00 INFO - [a1.1] -triggered off []
2021-07-29T17:19:07+12:00 INFO - [a2.1] -triggered off []
2021-07-29T17:19:15+12:00 INFO - [b1.1] -triggered off ['a1.1']
2021-07-29T17:19:24+12:00 INFO - [c.1] -triggered off ['a2.1', 'b1.1']
2021-07-29T17:19:33+12:00 INFO - [d1.1] -triggered off ['c.1']
```
On master all outputs are already optional, so we just need to infer `a1?succeed` etc. to get the same result with the new syntax.

Reply:
I think this example still relies on activating the suicide of `d2` to avoid getting partially satisfied dependencies.

Reply:
I remain unconvinced that we can reliably infer optional outputs based on suicide triggers.
Consider this Cylc 8 example:

```
a? => b1 => c1
a:fail? | a:submit-fail? => b2 => c2
c1 | c2 => d
```

At Cylc 7 this might be:
```
a => b1 => c1
a:fail | a:submit-fail => b2 => c2
c1 | c2 => d
d => !a & !b1 & !c1 & !b2 & !c2
```
At Cylc 7 there was nothing to stop users crudely suiciding a whole set of tasks (even if some of them completed successfully) rather than trying to work out some complicated logic to ensure a task only got suicided if necessary. Even if the optimal suicide triggers were used in this case I struggle to see how you would end up with the correct optional outputs.

So, I think we have to action suicide triggers when the pre-requisites are met and define what happens if a task is suicided (see the logic I proposed in #128 which may still be relevant?).
We know there are cases where we need this in any case for some Cylc 8 workflows.

I think we also have to treat Cylc 7 workflows (`suite.rc`) differently from Cylc 8 workflows (`flow.cylc`) for validation purposes.

Reply:
I think that case will work fine. We don't need to use suicide triggers to infer optional outputs because the presence of both `a:succeed` and `a:fail` implies success must be optional.  In addition to that, suicide triggers still work (by the spawn-on-demand logic that you commented on below). But if we infer optional outputs we won't end up with incomplete tasks - which would others indicate a serious error - that eventually get removed by suicide triggers.

Reply:
So the only difference for Cylc 7 suites is we encounter both `foo:succeed` and `foo:fail` we infer optionality rather than flagging it as an error.

Reply:
> So the only difference for Cylc 7 suites is if we encounter both foo:succeed and foo:fail we infer optionality rather than flagging it as an error.

OK - that's fine

> We don't need to use suicide triggers to infer optional outputs

Later in the proposal you say
"all we need to do is, once again, infer that x, y, and z are optional outputs (the suicide triggers imply this)".
So are you proposing to use suicide triggers to infer optional outputs in other cases?

Are we agreed that:
1) We will treat Cylc 7 workflows differently from Cylc 8 workflows for validation purposes?
2) We will identify Cylc 7 workflows  via the filename (`suite.rc`)?

Reply:
> Later in the proposal you say
"all we need to do is, once again, infer that x, y, and z are optional outputs (the suicide triggers imply this)".
So are you proposing to use suicide triggers to infer optional outputs in other cases?

Sorry if I wasn't sufficiently clear on this. It's true that the presence of suicide triggers implies there must be optional outputs involved, but in the succeed/fail case we don't need to use them to infer optional outputs. Also, the suicide triggers will still work if they need to.  In the succeed/fail case we can avoid the need though, by inferring optionality.

Branches off of custom outputs might be concurrent or alternate, because we don't know if the outputs are mutually exclusive or not. We could:
- try to infer optional cases from suicide triggers (might be tricky in some cases)
- leave them all as required outputs and let the existing suicide triggers clean up the resulting incomplete tasks

We can start with the second option, and consider trying the first later if it matters.

Reply:
> Are we agreed that:
> 
>     We will treat Cylc 7 workflows differently from Cylc 8 workflows for validation purposes?

Yes, we can't say use of both `foo:succeed` and `foo:fail` is an error in the Cylc 7 case.

>    We will identify Cylc 7 workflows via the filename (suite.rc)?
 
Presumably that's the easiest thing to do, but I haven't thought about that yet.